_posts/2023-04-30-writing-code-with-chatgpt-improve-it-with-kani.md

```diff
@@ -0,0 +1,359 @@
+For example, it does not compute the average for the input `integer_average(std::i32::MAX, 1)`.
+So, we tried again with the following prompt to generate a complete version of `integer_average`.
+
+```
+Q: Can you give me a fix that does not throw any error?
+```
+
+This time, ChatGPT found us an implementation that accepts all `i32` integers as input.
+
+```
+fn integer_average(num1: i32, num2: i32) -> i32 {
+    num1 / 2 + num2 / 2 + (num1 % 2 + num2 % 2) / 2
+}
+```
+
+Kani verifies that this implementation is correct.
+
+```
+> kani integer_average.rs
+
+...
+SUMMARY:
+ ** 0 of 6 failed
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 0.5309424s
+```
+
+#### Example 2.c
+Recall that we got a counterexample `x = 2147483647` from Kani for the equivalence check of `lowest_unset_bit`.
+We reported this counterexample to ChatGPT and asked for a fix.
+
+```
+Q: The optimized version does not produce the same value as the original for x = 2147483674.
+```
+
+It gave as a new optimization.
+
+``` rust
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    if x == std::u32::MAX{
+        return None;
+    }
+    Some((x ^ (x + 1)).trailing_zeros() as u8)
+} 
+```
+
+We ran Kani to check if the new version is equivalent to the original implementation with the same harness function.
+This time, there was only one failed check.
+
+```
+...
+SUMMARY:
+ ** 1 of 70 failed
+Failed Checks: assertion failed: lowest_unset_bit_ori(x) == lowest_unset_bit_opt(x)
+ File: "...", line 22, in check
+
+VERIFICATION:- FAILED
+Verification Time: 0.84979576s
+```
+
+The overflow has been fixed, but the equivalence failure persists. 
+Similarly, we produced another counterexample `x = 4293918719` with Kani, reported it to ChatGPT, and got another optimization.
+
+``` rust
+fn lowest_unset_bit(x: u32) -> Option<u8> {
+    if x == std::u32::MAX{
+        return None;
+    }
+    let y = !x;
+    Some(y.trailing_zeros() as u8)
+} 
+```
+
+Kani successfully verified that this optimization is equivalent to the original implementation.
+
+```
+> kani lowest_unset_bit.rs
+
+...
+SUMMARY:
+ ** 0 of 68 failed
+
+VERIFICATION:- SUCCESSFUL
+Verification Time: 5.308244s
+```
+
+The Kani proof saves us from spending a lot of time understanding the bit hack and convincing ourselves that the two implementations are equivalent.
+
+## Summary
+
+In this post, we showed how to generate programs with ChatGPT, how to verify generated code with Kani, and how to generate verified code with both ChatGPT and Kani.
+With increasingly powerful code generation capabilities, software development is becoming more automated and accessible.
+However, as developers spend less time on implementation details, corner cases, and safety issues, their code becomes more susceptible to errors.
+For this reason, specifications may become more critical than ever, serving as the new ground truth in cases where developers no longer write code line by line.
```

Comment:
Except for "must pass built-in checks" or "must be equivalent to this other program" the developed here have no specifications. Knowing what to verify on the generated code is the next big problem to address if the requirements are given only in natural language.